File: jarvis-personal/native/ios/DincrKit/Sources/DincrCore/FixtureDincrService.swift
import Foundation

/// In-memory service with synthetic data for SwiftUI previews, UI tests and store/landing
/// screenshots. All names and amounts are invented (CLAUDE.md §4.F: no real data in fixtures).
///
/// It stores what the screens write so flows can be exercised end to end, but it does not
/// compute financial results: dashboard figures are fixed sample values, exactly like the
/// backend would return them.
public actor FixtureDincrService: DincrService {
    public enum Scenario: String, Sendable {
        case populated, empty, failing, newUser
    }

    private var profile: Profile
    private var rows: [Movement]
    private let scenario: Scenario
    private let latency: Duration
    private var nextID = 100

    public init(scenario: Scenario = .populated, latency: Duration = .milliseconds(350), today: Date = .now) {
        self.scenario = scenario
        self.latency = latency
        self.profile = Profile(
            id: "fixture-account", email: "ana@example.com", displayName: "Ana Solís",
            profileSetupCompleted: scenario != .newUser,
            subscription: .init(plan: "free", status: "active")
        )
        self.rows = scenario == .populated ? Self.sampleMovements(today: today) : []
    }

    public func me() async throws -> Profile {
        try await pause()
        return profile
    }

    public func completeProfileSetup(_ setup: ProfileSetup) async throws -> Profile {
        try await pause()
        profile = Profile(
            id: profile.id, email: profile.email, displayName: setup.displayName,
            profileSetupCompleted: true, baseCurrency: setup.baseCurrency,
            numberFormat: setup.numberFormat, currencyPlacement: setup.currencyPlacement,
            subscription: profile.subscription
        )
        return profile
    }

    public func freeDashboard() async throws -> FreeDashboard {
        try await pause()
        if scenario == .empty || scenario == .newUser {
            return FreeDashboard(month: "2026-09", income: 0, expenses: 0, debtPaid: 0, debtBalance: nil, balance: 0,
                                 availableAfterCommitments: 0, categories: [], monthlyHistory: [])
        }
        return Self.sampleDashboard
    }

    public func movements() async throws -> [Movement] {
        try await pause()
        return rows.sorted { ($0.day ?? "") > ($1.day ?? "") }
    }

    public func create(_ kind: Movement.Kind, _ entry: EntryCreate) async throws {
        try await pause()
        nextID += 1
        let origin = kind == .income ? "salary" : "expense"
        rows.append(Movement(
            movementId: "\(origin):\(nextID)", sourceId: nextID, origin: origin, transactionDate: entry.entryDate,
            description: entry.description, amount: entry.amount, transactionType: kind, category: entry.category
        ))
    }

    public func update(movementID: String, _ update: MovementUpdate) async throws {
        try await pause()
        guard let index = rows.firstIndex(where: { $0.movementId == movementID }) else {
            throw APIError(kind: .notFound, status: 404, message: "Movimiento no encontrado o no editable.")
        }
        let old = rows[index]
        rows[index] = Movement(
            movementId: old.movementId, sourceId: old.sourceId, origin: old.origin, transactionDate: update.transactionDate,
            description: update.description, amount: update.amount, transactionType: old.transactionType,
            category: update.category, notes: update.notes, editable: old.editable
        )
    }

    public func delete(movementID: String) async throws {
        try await pause()
        rows.removeAll { $0.movementId == movementID }
    }

    private func pause() async throws {
        try await Task.sleep(for: latency)
        if scenario == .failing {
            throw APIError(kind: .server, status: 503, message: AppLanguage.current.pick(
                "DINCR no pudo completar la operación. Intentá de nuevo en unos segundos.",
                "DINCR couldn’t complete the operation. Try again in a few seconds."))
        }
    }

    static let sampleDashboard = FreeDashboard(
        month: "2026-09", income: 865_000, expenses: 512_450, debtPaid: 95_000, debtBalance: 1_240_000,
        balance: 257_550, availableAfterCommitments: 257_550,
        categories: [
            CategoryAmount(category: "Vivienda", amount: 210_000),
            CategoryAmount(category: "Comida", amount: 128_300),
            CategoryAmount(category: "Transporte", amount: 64_150),
            CategoryAmount(category: "Servicios", amount: 58_000),
            CategoryAmount(category: "Entretenimiento", amount: 32_000),
            CategoryAmount(category: "Salud", amount: 20_000),
        ],
        monthlyHistory: [
            MonthTotals(month: "2026-04", income: 820_000, expenses: 598_000, debtPaid: 95_000, balance: 127_000),
            MonthTotals(month: "2026-05", income: 820_000, expenses: 541_200, debtPaid: 95_000, balance: 183_800),
            MonthTotals(month: "2026-06", income: 905_000, expenses: 630_500, debtPaid: 95_000, balance: 179_500),
            MonthTotals(month: "2026-07", income: 840_000, expenses: 575_900, debtPaid: 95_000, balance: 169_100),
            MonthTotals(month: "2026-08", income: 865_000, expenses: 603_300, debtPaid: 95_000, balance: 166_700),
            MonthTotals(month: "2026-09", income: 865_000, expenses: 512_450, debtPaid: 95_000, balance: 257_550),
        ]
    )

    static func sampleMovements(today: Date) -> [Movement] {
        let calendar = Calendar(identifier: .gregorian)
        func day(_ offset: Int) -> String {
            let date = calendar.date(byAdding: .day, value: -offset, to: today) ?? today
            let parts = calendar.dateComponents([.year, .month, .day], from: date)
            return String(format: "%04d-%02d-%02d", parts.year!, parts.month!, parts.day!)
        }
        return [
            Movement(movementId: "expense:11", sourceId: 11, origin: "expense", transactionDate: day(0), description: "Supermercado", amount: 18_450, transactionType: .expense, category: "Comida"),
            Movement(movementId: "expense:10", sourceId: 10, origin: "expense", transactionDate: day(0), description: "Café", amount: 2_300, transactionType: .expense, category: "Restaurante"),
            Movement(movementId: "transaction:9", sourceId: 9, origin: "transaction", transactionDate: day(1), description: "Aviso bancario · Gasolinera", amount: 25_000, transactionType: .expense, category: "Gasolina", editable: false),
            Movement(movementId: "salary:8", sourceId: 8, origin: "salary", transactionDate: day(3), description: "Salario quincenal", amount: 432_500, transactionType: .income, category: "Salario"),
            Movement(movementId: "expense:7", sourceId: 7, origin: "expense", transactionDate: day(4), description: "Internet del hogar", amount: 24_900, transactionType: .expense, category: "Internet"),
            Movement(movementId: "expense:6", sourceId: 6, origin: "expense", transactionDate: day(6), description: "Farmacia", amount: 9_800, transactionType: .expense, category: "Salud"),
            Movement(movementId: "expense:5", sourceId: 5, origin: "expense", transactionDate: day(9), description: "Alquiler", amount: 210_000, transactionType: .expense, category: "Vivienda"),
            Movement(movementId: "salary:4", sourceId: 4, origin: "salary", transactionDate: day(18), description: "Salario quincenal", amount: 432_500, transactionType: .income, category: "Salario"),
        ]
    }
}

File: jarvis-personal/native/android/core/data/src/main/kotlin/com/dincr/data/FixtureDincrService.kt
package com.dincr.data

import java.math.BigDecimal
import java.time.LocalDate
import kotlinx.coroutines.delay
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

/**
 * In-memory service with synthetic data for previews, UI tests and screenshots. All names and
 * amounts are invented (CLAUDE.md §4.F). Dashboard figures are fixed sample values, exactly as
 * the backend would return them; nothing is computed here. Write semantics follow the backend:
 * a repeated idempotency key adds no second row, and a missing movement is a 404.
 */
class FixtureDincrService(
    private val scenario: Scenario = Scenario.POPULATED,
    private val latencyMs: Long = 350,
    today: LocalDate = LocalDate.now(),
) : DincrService {
    enum class Scenario { POPULATED, EMPTY, FAILING, NEW_USER }

    private val mutex = Mutex()
    private var profile = Profile(
        // Every field /auth/me always sends (auth/saas.py enrich_identity), same as the iOS fixture.
        id = 4201, email = "ana@example.com", displayName = "Ana Solís", role = "user", planSelected = true,
        baseCurrency = "CRC", numberFormat = "dot_comma", currencyPlacement = "before",
        profileSetupCompleted = scenario != Scenario.NEW_USER,
        subscription = Profile.Subscription(plan = "free", status = "active"),
    )
    private val rows = if (scenario == Scenario.POPULATED) sampleMovements(today).toMutableList() else mutableListOf()
    private var nextId = 100
    private val seenKeys = mutableSetOf<String>()

    override suspend fun me(): Profile = pause { profile }

    override suspend fun completeProfileSetup(setup: ProfileSetup): Profile = pause {
        profile = profile.copy(
            displayName = setup.displayName, profileSetupCompleted = true, baseCurrency = setup.baseCurrency,
            numberFormat = setup.numberFormat, currencyPlacement = setup.currencyPlacement,
        )
        profile
    }

    override suspend fun freeDashboard(): FreeDashboard = pause {
        if (scenario == Scenario.EMPTY || scenario == Scenario.NEW_USER) {
            // The backend always returns six months, zero-filled.
            val zero = BigDecimal.ZERO
            FreeDashboard(month = "2026-09", income = zero, expenses = zero, debtPaid = zero, debtBalance = zero, balance = zero, availableAfterCommitments = zero,
                monthlyHistory = listOf("2026-04", "2026-05", "2026-06", "2026-07", "2026-08", "2026-09").map { MonthTotals(it, zero, zero, zero, zero) })
        } else SAMPLE_DASHBOARD
    }

    override suspend fun movements(): List<Movement> = pause { rows.sortedByDescending { it.day.orEmpty() } }

    override suspend fun create(kind: MovementKind, entry: EntryCreate, idempotencyKey: String) = pause {
        if (!seenKeys.add(idempotencyKey)) return@pause
        nextId += 1
        val origin = if (kind == MovementKind.INCOME) "salary" else "expense"
        rows += Movement(
            movementId = "$origin:$nextId", sourceId = nextId, origin = origin, transactionDate = entry.entryDate,
            description = entry.description, amount = entry.amount,
            transactionType = if (kind == MovementKind.INCOME) "income" else "expense", category = entry.category, editable = true,
        )
    }

    override suspend fun update(movementId: String, update: MovementUpdate) = pause {
        val index = rows.indexOfFirst { it.movementId == movementId }
        if (index < 0) throw ApiError(ApiError.Kind.NOT_FOUND, 404, message = "Movimiento no encontrado o no editable.")
        rows[index] = rows[index].copy(transactionDate = update.transactionDate, description = update.description, amount = update.amount, category = update.category, notes = update.notes)
    }

    override suspend fun delete(movementId: String) = pause {
        if (!rows.removeAll { it.movementId == movementId }) throw ApiError(ApiError.Kind.NOT_FOUND, 404, message = "Movimiento no encontrado o no eliminable.")
    }

    private suspend fun <T> pause(block: () -> T): T {
        if (latencyMs > 0) delay(latencyMs)
        if (scenario == Scenario.FAILING) {
            throw ApiError(ApiError.Kind.SERVER, 503, message = AppLanguage.current().pick(
                "DINCR no pudo completar la operación. Intentá de nuevo en unos segundos.",
                "DINCR couldn’t complete the operation. Try again in a few seconds."))
        }
        return mutex.withLock { block() }
    }

    companion object {
        private fun m(value: Long) = BigDecimal.valueOf(value)

        val SAMPLE_DASHBOARD = FreeDashboard(
            month = "2026-09", income = m(865_000), expenses = m(512_450), debtPaid = m(95_000), debtBalance = m(1_240_000),
            balance = m(257_550), availableAfterCommitments = m(257_550),
            categories = listOf(
                CategoryAmount("Vivienda", m(210_000)), CategoryAmount("Comida", m(128_300)), CategoryAmount("Transporte", m(64_150)),
                CategoryAmount("Servicios", m(58_000)), CategoryAmount("Entretenimiento", m(32_000)), CategoryAmount("Salud", m(20_000)),
            ),
            monthlyHistory = listOf(
                MonthTotals("2026-04", m(820_000), m(598_000), m(95_000), m(127_000)),
                MonthTotals("2026-05", m(820_000), m(541_200), m(95_000), m(183_800)),
                MonthTotals("2026-06", m(905_000), m(630_500), m(95_000), m(179_500)),
                MonthTotals("2026-07", m(840_000), m(575_900), m(95_000), m(169_100)),
                MonthTotals("2026-08", m(865_000), m(603_300), m(95_000), m(166_700)),
                MonthTotals("2026-09", m(865_000), m(512_450), m(95_000), m(257_550)),
            ),
        )

        fun sampleMovements(today: LocalDate): List<Movement> {
            fun day(offset: Long) = today.minusDays(offset).toString()
            fun row(id: String, offset: Long, text: String, amount: BigDecimal, type: String, category: String, editable: Boolean = true) =
                Movement(movementId = id, sourceId = id.substringAfter(':').toInt(), origin = id.substringBefore(':'), transactionDate = day(offset),
                    description = text, amount = amount, transactionType = type, category = category, editable = editable)
            fun row(id: String, offset: Long, text: String, amount: Long, type: String, category: String, editable: Boolean = true) =
                row(id, offset, text, m(amount), type, category, editable)
            return listOf(
                row("expense:11", 0, "Supermercado", 18_450, "expense", "Comida"),
                row("expense:10", 0, "Café", 2_300, "expense", "Restaurante"),
                // Cents and a category outside the editor's list: editing must keep both.
                row("expense:12", 1, "Feria del agricultor", BigDecimal("12345.5"), "expense", "Feria"),
                row("transaction:9", 1, "Aviso bancario · Gasolinera", 25_000, "expense", "Gasolina", editable = false),
                row("salary:8", 3, "Salario quincenal", 432_500, "income", "Salario"),
                row("expense:7", 4, "Internet del hogar", 24_900, "expense", "Internet"),
                row("expense:6", 6, "Farmacia", 9_800, "expense", "Salud"),
                row("expense:5", 9, "Alquiler", 210_000, "expense", "Vivienda"),
                row("salary:4", 18, "Salario quincenal", 432_500, "income", "Salario"),
            )
        }
    }
}
